### PR TITLE
Implement HttpContent and HttpClient cancellation APIs

### LINKED ISSUE #576: More cancellation overloads for HttpContent
Part of a series of issues adding cancellation support to HttpContent and HttpClient:
https://github.com/dotnet/corefx/issues/32615, https://github.com/dotnet/corefx/issues/9071

In order to implement `HttpContent.ReadAsStreamAsync(CancellationToken)` from dotnet/corefx#32615, an additional overload for `CreateContentReadStreamAsync` is needed (see [comment](https://github.com/dotnet/corefx/issues/32615#issuecomment-562083237)).

The default implementation would drop the `CancellationToken` and call the existing overload. This is the behavior that was decided for `SerializeToStreamAsync` in dotnet/corefx#9071.

</br>

`HttpContent` also lacks support for canceling `CopyToAsync`.

</br>

Proposal:

```c#
partial class HttpContent
{
    // Existing
    protected virtual Task<Stream> CreateContentReadStreamAsync();
    // New
    protected virtual Task<Stream> CreateContentReadStreamAsync(CancellationToken cancellationToken);


    // Existing
    public Task CopyToAsync(Stream stream);
    public Task CopyToAsync(Stream stream, TransportContext context);
    internal Task CopyToAsync(Stream stream, TransportContext context, CancellationToken cancellationToken);

    // New
    public Task CopyToAsync(Stream stream, CancellationToken cancellationToken);
    // Change internal => public
    public Task CopyToAsync(Stream stream, TransportContext context, CancellationToken cancellationToken);
}
```

### LINKED ISSUE #916: Add cancellation overloads on HttpClient and HttpContent
It seems obvious that there is a lack of external cancellation support in a few HttpClient methods: GetByteArrayAsync, GetStreamAsync and maybe others. Any particular reason?

What is expected:

```c#
class HttpClient {
    public Task<byte[]> GetByteArrayAsync(string requestUri, CancellationToken token = default);
    public Task<byte[]> GetByteArrayAsync(Uri requestUri, CancellationToken token = default);

    public Task<byte[]> GetStreamAsync(string requestUri, CancellationToken token = default);
    public Task<byte[]> GetStreamAsync(Uri requestUri, CancellationToken token = default);

    public Task<byte[]> GetStringAsync(string requestUri, CancellationToken token = default);
    public Task<byte[]> GetStringAsync(Uri requestUri, CancellationToken token = default);
}
class HttpContent {
    public Task<byte[]> GetByteArrayAsync(CancellationToken token = default);
    public Task<byte[]> GetStreamAsync(CancellationToken token = default);
    public Task<byte[]> GetStringAsync(CancellationToken token = default);
}
```

I might be not alone here. If missed another case or the feature is in the pipeline already, then comment to the case and close as a duplicate.

Thanks. 

P.S.
`GetByteArrayAsync` is another overload candidate mentioned by dotnet/corefx#32762.

### LINKED ISSUE #17492: Add HttpContent.SerializeToStreamAsync overload with cancellation
We should consider adding overloads for methods like SerializeToStreamAsync, so that cancellation can be passed in and threaded through to the underlying ReadAsync/WriteAsync operations performed.

Note: Read* methods are now covered in dotnet/runtime#916
```c#
class HttpContent {
    protected virtual Task SerializeToStreamAsync(Stream stream, TransportContext context, CancellationToken token);
    // EXISTING API:
    protected abstract Task SerializeToStreamAsync(Stream stream, TransportContext context);
}
```

### PR DESCRIPTION
**Pull request by MihaZupan**
Closes #916 (dotnet/corefx#32615), closes dotnet/runtime#576 and closes #17492 (dotnet/corefx#9071).

This PR implements the following APIs:
```c#
partial class HttpClient
{
    public Task<byte[]> GetByteArrayAsync(string requestUri, CancellationToken cancellationToken);
    public Task<byte[]> GetByteArrayAsync(Uri requestUri, CancellationToken cancellationToken);

    public Task<Stream> GetStreamAsync(string requestUri, CancellationToken cancellationToken);
    public Task<Stream> GetStreamAsync(Uri requestUri, CancellationToken cancellationToken);

    public Task<string> GetStringAsync(string requestUri, CancellationToken cancellationToken);
    public Task<string> GetStringAsync(Uri requestUri, CancellationToken cancellationToken);
}

partial class HttpContent
{
    public Task<byte[]> ReadAsByteArrayAsync(CancellationToken cancellationToken);
    public Task<Stream> ReadAsStreamAsync(CancellationToken cancellationToken);
    public Task<string> ReadAsStringAsync(CancellationToken cancellationToken);

    protected virtual Task<Stream> CreateContentReadStreamAsync(CancellationToken cancellationToken);

    protected virtual Task SerializeToStreamAsync(Stream stream, TransportContext context, CancellationToken cancellationToken);

    public Task CopyToAsync(Stream stream, CancellationToken cancellationToken);
    public Task CopyToAsync(Stream stream, TransportContext context, CancellationToken cancellationToken);
}
```

### CODE REVIEW

**Comment by stephentoub** [comment]
> Blocked by #576, awaiting approval.

We prefer to not open PRs for new APIs until the the APIs have been approved.

**Comment by MihaZupan** [comment]
> We prefer to not open PRs for new APIs until the the APIs have been approved.

I was advised that I should mark the PR as blocked in the case it's only adding CTs to existing overloads.

>  Can you make the relevant APIs internal until that issue is reviewed/approved? That way the ones that have been approved can proceed in this PR.

They are be up for review tomorrow, I assume the PR review will take longer than that, otherwise I can make them internal as you suggested.

**Comment by stephentoub** [comment]
> I was advised that I should mark the PR as blocked

Thanks. By whom?

**Comment by MihaZupan** [comment]
By @karelz during the last sync meeting. That was only referring to the need of 

```c#
protected virtual Task<Stream> CreateContentReadStreamAsync(CancellationToken cancellationToken);
```

as described in https://github.com/dotnet/corefx/issues/32615#issuecomment-562083237.

Cory asked me to add the `CopyToAsync` overloads to the same API proposal issue, but I don't think those should be a contentious topic.

File: src/libraries/Common/src/System/Net/Http/NoWriteNoSeekStreamContent.cs
Comment (davidsh) [inline src/libraries/Common/src/System/Net/Http/NoWriteNoSeekStreamContent.cs]: We shouldn't need the "HTTP_DLL" define anymore. It was used only when we were building WinHttpHandler as part of System.Net.Http.dll. But @ManickaP PR got rid of building WinHttpHandler as part of System.Net.Http.dll.

So, you should remove "HTTP_DLL" define from this file and any other place it might appear such as CSPROJ files, etc.

Comment (MihaZupan) [reply src/libraries/Common/src/System/Net/Http/NoWriteNoSeekStreamContent.cs]: As far as I can tell WinHttpHandler and System.Net.Http link to NoWriteNoSeekStreamContent.cs. 

I've been fighting with the build for the past couple hours regarding this as it seems the netfx version of WinHttpHandler targets a different definition of HttpContent and therefore requires a different signature for the changed method (SerializeToStreamAsync).

Comment (davidsh) [reply src/libraries/Common/src/System/Net/Http/NoWriteNoSeekStreamContent.cs]: >I've been fighting with the build for the past couple hours regarding this as it seems the netfx version of WinHttpHandler targets a different definition of HttpContent and therefore requires a different signature for the changed method (SerializeToStreamAsync).

Yes, WinHttpHandler doesn't build against 'netcoreapp'. It builds against 'netstandard20' which won't have the new APIs.

You might need to add new defines to work around this problem.

Comment (ManickaP) [reply src/libraries/Common/src/System/Net/Http/NoWriteNoSeekStreamContent.cs]: I think I get it, but please correct me @davidsh if I'm wrong. The HTTP_DLL should be removed, but very similar constant must be introduced for WinHttpHandler library with pretty much the opposite meaning to HTTP_DLL. Then filter overrides and access modifier according to this.

Comment (davidsh) [reply src/libraries/Common/src/System/Net/Http/NoWriteNoSeekStreamContent.cs]: Correct.

We are now introducing new APIs in .NET Core which affects the building of the WinHttpHandler library (the new overloads with cancellationtoken). So we need to use compile-time mechanisms to preserve the ability to build WinHttpHandler against 'netstandard20' targets (which include .NET Framework). And 'netstandard20' target doesn't have those new APIs available.

We have two different ways to achieve compile-time differences. One is thru the use of these defines. The other is partial classes and separating out into different files (with suffixes like .netcoreapp.cs or .netfx.cs etc.) like we did for other Http stuff.  But when the code is scattered like this it is easier to use defines similar to 'HTTP_DLL'.

Comment (MihaZupan) [reply src/libraries/Common/src/System/Net/Http/NoWriteNoSeekStreamContent.cs]: Considering only this one class is problematic in this case (with a single 'real' method), would it make sense to duplicate the file for System.Net.Http and the WinHttpHandler?

Comment (davidsh) [reply src/libraries/Common/src/System/Net/Http/NoWriteNoSeekStreamContent.cs]: >would it make sense to duplicate the file for System.Net.Http and the WinHttpHandler

No. We don't like to duplicate any sources. That's why we put files into Common folders etc.

For small things like this, it is easier to just use compile-time define constants.

File: src/libraries/Common/src/System/Net/Http/NoWriteNoSeekStreamContent.cs
Comment (davidsh) [inline src/libraries/Common/src/System/Net/Http/NoWriteNoSeekStreamContent.cs]: I would prefer this be called "WINHTTPHANDLER" or "WINHTTP_HANDLER".  Leaving it with 3 separate words is confusing since it won't make it obvious that this is about the System.Net.Http.WinHttpHandler.dll binary and related NuGet package.

Comment (MihaZupan) [reply src/libraries/Common/src/System/Net/Http/NoWriteNoSeekStreamContent.cs]: I personally find it more readable with every word separated but don't have a strong preference. Would WINHTTPHANDLER_DLL work to emphasize that this is a different assembly and not just a random switch?

Comment (davidsh) [reply src/libraries/Common/src/System/Net/Http/NoWriteNoSeekStreamContent.cs]: >Would WINHTTPHANDLER_DLL work to emphasize that this is a different assembly and not just a random switch?

That would work.

File: src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs
Comment (ManickaP) [inline src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: Thank you for cleaning this up!

Comment (ManickaP) [inline src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: Thank you for cleaning this up!

File: src/libraries/Common/src/System/Net/Http/NoWriteNoSeekStreamContent.cs
Comment (davidsh) [reply src/libraries/Common/src/System/Net/Http/NoWriteNoSeekStreamContent.cs]: Are you still planning on changing that define to 'WINHTTPHANDLER_DLL' ?

File: src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs
Comment (stephentoub) [inline src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: If we wanted to we could add a configuration to the System.Net.Http.WinHttpHandler package that would contain a build for .NET 5, and then we could utilize the new APIs in that build.  But maybe it's not worth it.

File: src/libraries/Common/src/System/Net/Http/NoWriteNoSeekStreamContent.cs
Comment (MihaZupan) [reply src/libraries/Common/src/System/Net/Http/NoWriteNoSeekStreamContent.cs]: Unless there are other suggestions, yes

File: src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs
Comment (davidsh) [reply src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: It would only help in scenarios where someone wanted to cancel an operation that uploaded request body content (i.e. POST).

I think we can wait for customer feedback and then decide if it is worth adding that ifdef'd code and additional .NET 5 build for the Nuget package.

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientTest.cs
Comment (davidsh) [inline src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientTest.cs]: nit: In cases where you are doing a `new`, you can use `var` instead of the type name, i.e.

```c#
var onClientFinished = new SemaphoreSlim(0, 1);
```

Comment applies to other places with similar pattern.

The use of `var` is allowed when the type is obvious such as a `new` or an explicit cast.

File: src/libraries/System.Net.Http/src/System/Net/Http/HttpContent.cs
Comment (scalablecory) [inline src/libraries/System.Net.Http/src/System/Net/Http/HttpContent.cs]: I don't think the usability win would be huge here, but we should consider having the default implementation of both methods check to see if either was overloaded and forward to the right one.

Comment (MihaZupan) [reply src/libraries/System.Net.Http/src/System/Net/Http/HttpContent.cs]: All BCL http content implementations that can make use of cancellation already do that. Stephen added most, i added that on MultipartContent in this PR

**Review by davidsh — APPROVED** [review]

**Review by scalablecory — APPROVED** [review]

**Pull request merged**
